vm/qemu/snapshot_linux.go

```diff
@@ -0,0 +1,248 @@
+		return fmt.Errorf("qemu: ivs conn sendmsg failed: %w", err)
+	}
+	// Send event FD for VM 1 interrupt vector 0.
+	binary.LittleEndian.PutUint64(msg, 1)
+	rights = syscall.UnixRights(inst.eventFD)
+	if _, _, err := conn.WriteMsgUnix(msg, rights, nil); err != nil {
+		return fmt.Errorf("qemu: ivs conn sendmsg failed: %w", err)
+	}
+	return nil
+}
+
+func (inst *instance) SetupSnapshot(input []byte) error {
+	copy(inst.input, input)
+	// Tell executor that we are ready to snapshot and wait for an ack.
+	inst.setSnapshotState(flatrpc.SnapshotStateHandshake)
+	if !inst.waitSnapshotStateChange(flatrpc.SnapshotStateHandshake, 10*time.Minute) {
+		return fmt.Errorf("executor does not start snapshot handshake")
+	}
+	if _, err := inst.hmp("migrate_set_capability x-ignore-shared on", 0); err != nil {
+		return err
+	}
+	if _, err := inst.hmp("savevm syz", 0); err != nil {
+		return err
+	}
+	if inst.debug {
+		inst.hmp("info snapshots", 0) // this prints size of the snapshot
+	}
+	inst.setSnapshotState(flatrpc.SnapshotStateSnapshotted)
+	if !inst.waitSnapshotStateChange(flatrpc.SnapshotStateSnapshotted, time.Minute) {
+		return fmt.Errorf("executor has not confirmed snapshot handshake")
+	}
+	return nil
+}
+
+func (inst *instance) RunSnapshot(timeout time.Duration, input []byte) (result, output []byte, err error) {
+	copy(inst.input, input)
+	inst.header.OutputOffset = 0
+	inst.header.OutputSize = 0
+	inst.setSnapshotState(flatrpc.SnapshotStateExecute)
+	if _, err := inst.hmp("loadvm syz", 0); err != nil {
+		return nil, nil, err
+	}
+	inst.waitSnapshotStateChange(flatrpc.SnapshotStateExecute, timeout)
+	resStart := int(flatrpc.ConstMaxInputSize) + int(atomic.LoadUint32(&inst.header.OutputOffset))
+	resEnd := resStart + int(atomic.LoadUint32(&inst.header.OutputSize))
+	var res []byte
+	if resEnd <= len(inst.shmem) {
+		res = inst.shmem[resStart:resEnd:resEnd]
+	}
+	output = inst.readOutput()
+	return res, output, nil
+}
+
+func (inst *instance) setSnapshotState(state flatrpc.SnapshotState) {
+	inst.header.State = state
+}
+
+func (inst *instance) waitSnapshotStateChange(state flatrpc.SnapshotState, timeout time.Duration) bool {
+	deadline := time.Now().Add(timeout)
+	timeoutMs := int(timeout / time.Millisecond)
+	fds := []unix.PollFd{{
+		Fd:     int32(inst.eventFD),
+		Events: unix.POLLIN,
+	}}
+	for {
+		if n, _ := unix.Poll(fds, timeoutMs); n == 1 {
+			var buf [8]byte
+			syscall.Read(inst.eventFD, buf[:])
+		}
+		if inst.header.State != state {
+			return true
+		}
+		remain := time.Until(deadline)
+		if remain < time.Millisecond {
+			return false
+		}
+		timeoutMs = int(remain / time.Millisecond)
+	}
+}
+
+func (inst *instance) readOutput() []byte {
+	var output []byte
+	// If output channel has overflown, then wait for more output.
+	wait := cap(inst.merger.Output)
+	for {
+		select {
+		case out := <-inst.merger.Output:
+			output = append(output, out...)
+			wait--
+		default:
+			if wait > 0 {
```

Comment:
It's not immediately clear what happens here.

Reply:
Added a comment.